src/ucs/datastruct/ptr_array.c

```diff
@@ -174,26 +195,40 @@ unsigned ucs_ptr_array_insert(ucs_ptr_array_t *ptr_array, void *value,
     /* Get the first item on the free list */
     index = ptr_array->freelist;
     ucs_assert(index != UCS_PTR_ARRAY_SENTINEL);
-    elem = &ptr_array->start[index];
 
-    /* Remove from free list */
-    ptr_array->freelist = ucs_ptr_array_freelist_get_next(*elem);
+    elem = &ptr_array->start[index];
 
-    /* Populate */
-    *placeholder_p = ucs_ptr_array_placeholder_get(*elem);
+    /* Remove from free list and Populate */
+    ptr_array->freelist = ucs_ptr_array_freelist_get_next_params(*elem,
+        &size_free_ahead);
```

Comment:
Don't you need to set the "free_size" on the next free element? Otherwise, you need to initialize all the free elements in each chunk to have the correct "free_size" ahead of them all the time, no?

Reply:
The new method works as follows:
On remove of an index: If start[index+1] is free ==> 
   start[index].free_elements_ahead = start[index+1].free_elements_ahead+1
   Then, the removed index is pushed to the HEAD of the freelist.
   NOTE, that if start[index+1] is free ==> It's already in the freelist !!!

On insert, the we fetch the first entry of the freelist and we rely on the fact that 
this remove/insert is effectively a LIFO, i.e. the last item pushed into the freelist will
be fetched first ==> There is no chance that index+1 will be fetched before index since index+1 was already in the list before index was put into the list.
Therefore, we can rely on the free_size_ahead field to tell how many free elements are from any index in the freelist.

Reply:
Added an explanation in ptr_array.h:
/*
 * Array element layout:
 *
 *         64                 32               1   0
 *          +-----------------+----------------+---+
 * free:    |     size_free   |  next index    | 1 |
 *          +-----------------+----------------+---+
 * used:    |           user pointer           | 0 |
 *          +-----------------+----------------+---+
 *
 *
 * size_free indicates the number of consecutive free elements ahead.
 *
 * The remove / insert algorithm works as follows:
 * On remove of an index: If start[index+1] is free ==> 
 *    start[index].free_elements_ahead = start[index+1].free_elements_ahead+1
 *  Then, the removed index is pushed to the HEAD of the freelist.
 *  NOTE, that if start[index+1] is free ==> It's already in the freelist !!!
 *
 * On insert, the we fetch the first entry of the freelist and we rely on the fact that 
 * the remove/insert mechanism effectively implements a LIFO freelist, i.e. the last
 * item pushed into the freelist will be fetched first ==> There is no chance that 
 * index+1 will be fetched before index since index+1
 * was already in the list before index was put into the list.
 *
 * Therefore, we can rely on the free_size_ahead field to tell how many free
 * elements are from any index in the freelist.
 *
 */